tiledb/sm/filesystem/vfs.h

```diff
@@ -394,6 +394,14 @@ class VFS : FilesystemBase,
   /*               API                 */
   /* ********************************* */
 
+  /**
+   * Returns the filesystem which corresponds to the given URI.
+   *
+   * @param uri The URI which may correspond to a filesystem.
+   * @return The filesystem on which `uri` is stored.
+   */
+  const FilesystemBase& get_fs(const URI& uri) const;
+
```

Comment:
```suggestion

  FilesystemBase& get_fs(const URI uri) {
    return const_cast<FilesystemBase&>(
        static_cast<const VFS*>(this)->get_fs(uri));
  }
```

With this I think you can remove the `const_cast`s throughout `vfs.cc`.